x-pack/filebeat/input/awss3/sqs_s3_event.go

```diff
@@ -152,15 +167,13 @@ func (p *sqsS3EventProcessor) ProcessSQS(ctx context.Context, msg *types.Message
 		return nil
 	}
 
-	if p.maxReceiveCount > 0 && !errors.Is(processingErr, &nonRetryableError{}) {
+	if p.maxReceiveCount > 0 && !errors.Is(processingErr, &nonRetryableError{}) && hasReceiveCountAttribute {
 		// Prevent poison pill messages from consuming all workers. Check how
 		// many times this message has been received before making a disposition.
-		if v, found := msg.Attributes[sqsApproximateReceiveCountAttribute]; found {
-			if receiveCount, err := strconv.Atoi(v); err == nil && receiveCount >= p.maxReceiveCount {
-				processingErr = nonRetryableErrorWrap(fmt.Errorf(
-					"sqs ApproximateReceiveCount <%v> exceeds threshold %v: %w",
-					receiveCount, p.maxReceiveCount, processingErr))
-			}
+		if receiveConvErr == nil && receiveCount >= p.maxReceiveCount {
+			processingErr = nonRetryableErrorWrap(fmt.Errorf(
+				"sqs ApproximateReceiveCount <%v> exceeds threshold %v: %w",
+				receiveCount, p.maxReceiveCount, processingErr))
 		}
 	}
 
```

Comment:
What if we add a helper function to get the receive count? This is simpler to read IMO. Might even get better if you add a helper for `getSQSSentTimestamp`.

```diff
diff --git a/x-pack/filebeat/input/awss3/sqs_s3_event.go b/x-pack/filebeat/input/awss3/sqs_s3_event.go
index e37b6254c4..0ab0e3376e 100644
--- a/x-pack/filebeat/input/awss3/sqs_s3_event.go
+++ b/x-pack/filebeat/input/awss3/sqs_s3_event.go
@@ -134,12 +134,10 @@ func (p *sqsS3EventProcessor) ProcessSQS(ctx context.Context, msg *types.Message
 	keepaliveWg.Add(1)
 	go p.keepalive(keepaliveCtx, log, &keepaliveWg, msg)
 
-	rawRecieveCount, hasReceiveCountAttribute := msg.Attributes[sqsApproximateReceiveCountAttribute]
-	receiveCount, receiveConvErr := strconv.Atoi(rawRecieveCount)
-
-	// Only contribute to the sqs_lag_time histogram on the first message
-	// to avoid skewing the metric when processing retries
-	if hasReceiveCountAttribute && receiveConvErr == nil && receiveCount == 1 {
+	receiveCount := getSQSReceiveCount(msg.Attributes)
+	if receiveCount == 1 {
+		// Only contribute to the sqs_lag_time histogram on the first message
+		// to avoid skewing the metric when processing retries.
 		if s, found := msg.Attributes[sqsSentTimestampAttribute]; found {
 			if sentTimeMillis, err := strconv.ParseInt(s, 10, 64); err == nil {
 				sentTime := time.UnixMilli(sentTimeMillis)
@@ -167,10 +165,10 @@ func (p *sqsS3EventProcessor) ProcessSQS(ctx context.Context, msg *types.Message
 		return nil
 	}
 
-	if p.maxReceiveCount > 0 && !errors.Is(processingErr, &nonRetryableError{}) && hasReceiveCountAttribute {
+	if p.maxReceiveCount > 0 && !errors.Is(processingErr, &nonRetryableError{}) {
 		// Prevent poison pill messages from consuming all workers. Check how
 		// many times this message has been received before making a disposition.
-		if receiveConvErr == nil && receiveCount >= p.maxReceiveCount {
+		if receiveCount >= p.maxReceiveCount {
 			processingErr = nonRetryableErrorWrap(fmt.Errorf(
 				"sqs ApproximateReceiveCount <%v> exceeds threshold %v: %w",
 				receiveCount, p.maxReceiveCount, processingErr))
@@ -363,3 +361,14 @@ func (p *sqsS3EventProcessor) finalizeS3Objects(handles []s3ObjectHandler) error
 	}
 	return multierr.Combine(errs...)
 }
+
+// getSQSReceiveCount returns the SQS ApproximateReceiveCount attribute. If the value
+// cannot be read then -1 is returned.
+func getSQSReceiveCount(attributes map[string]string) int {
+	if s, found := attributes[sqsApproximateReceiveCountAttribute]; found {
+		if receiveCount, err := strconv.Atoi(s); err == nil {
+			return receiveCount
+		}
+	}
+	return -1
+}
```

(If you turn that ^ into a file then you can `git apply <file>` to incorporate this into your branch.)